karaf/osgi-jta/src/main/java/org/jboss/narayana/quickstart/osgi/RecoveryExample.java

```diff
@@ -0,0 +1,160 @@
+import org.jboss.tm.XAResourceRecovery;
+import org.osgi.framework.BundleContext;
+
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.sql.DataSource;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import java.io.File;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+/**
+ * @author <a href="mailto:zfeng@redhat.com">Amos Feng</a>
+ */
+
+@Command(scope = "narayana-quickstart", name = "testRecovery", description="The JTA Recovery Example")
+@Service
+public class RecoveryExample implements Action {
+    @Reference
+    private BundleContext context;
+
+    @Reference
+    private TransactionManager tm;
+
+    @Option(name = "-f", aliases = "--force", description = "force the example to crash", required = false, multiValued = false)
+    boolean force = false;
+
+    @Override
+    public Object execute() throws Exception {
+        if (force) {
+            System.out.println("testRecovery generate something to recovery");
+            if (tm == null) {
+                System.out.println("Can not get transaction manager");
+            } else {
+                DummyXAResource xa = new DummyXAResource(DummyXAResource.faultType.HALT);
+                Context ctx = new InitialContext();
+                DataSource ds = (DataSource)ctx.lookup("osgi:service/test");
+                Connection conn = null;
+                Statement stmt = null;
+
+                try {
+                    tm.begin();
+                    Transaction transaction = tm.getTransaction();
+                    transaction.enlistResource(xa);
+
+                    conn = ds.getConnection();
+                    stmt = conn.createStatement();
+                    stmt.execute("CREATE TABLE IF NOT EXISTS example ( id varchar(10), name varchar(32) )");
+                    stmt.execute("insert into example values (1, 'recovery')");
+
+                    tm.commit();
+                } catch (SQLException e) {
+                    System.out.println(e);
+                    tm.rollback();
+                } finally {
+                    stmt.close();
+                    conn.close();
+                }
+            }
+        } else {
+            File file = new File("DummyXAResource/");
+            if (file.exists() && file.isDirectory() && file.listFiles().length > 0) {
+                System.out.println("register the DummyXAResourceRecovery");
+                int currentCommits = DummyXAResource.getCommitRequests();
+                context.registerService(XAResourceRecovery.class, new DummyXAResourceRecovery(), null);
+
+                System.out.println("query before recovery");
+                queryH2Database("select * from example");
+
+                System.out.println("testRecovery waiting ...");
+                while (DummyXAResource.getCommitRequests() == currentCommits) {
```

Comment:
Can I just ask, I don't see synchronization between this and the recovery manager. Is it possible that if this part of the test is slow to run (stuck in a debugger) then it would look to pass but not actually do so?

I think you need to start the recovery manager _after_ you have queried H2 first time. I think you need to also put something to assert that after recovery commitRequests = currentCommits+1

Maybe I misread and if I did it would be great to get some comments in here to explain what is happening and what I should be seeing.